lib/school_house/content/lesson.ex

```diff
@@ -68,42 +69,32 @@ defmodule SchoolHouse.Content.Lesson do
     end)
   end
 
-  defp table_of_contents_reducer([_, "h" <> size, name], {html, header}) do
-    section_link =
-      name
-      |> String.trim()
-      |> page_link()
-
-    size = String.to_integer(size)
+  defp build_table(matches, acc \\ [])
 
-    cond do
-      is_nil(header) ->
-        {"#{html}<li>#{section_link}", size}
+  defp build_table([], []) do
+    ""
+  end
 
-      size == header ->
-        {"#{html}</li><li>#{section_link}", size}
+  defp build_table([], acc) do
+    "<ul>#{acc |> Enum.reverse() |> Enum.join("")}</ul>"
+  end
 
-      size < header ->
-        closing_tag =
-          1..(header - size)
-          |> Enum.map(fn _ -> "</ul>" end)
-          |> Enum.join("")
+  defp build_table([{size, name} | matches], acc) do
+    {children, remaining} = Enum.split_while(matches, fn {s, _} -> s > size end)
+    children = build_table(children)
 
-        {"#{html}</li>#{closing_tag}</li><li>#{section_link}", size}
+    section_link =
+      name
+      |> String.trim()
+      |> page_link()
 
-      size > header ->
-        {"#{html}<ul><li>#{section_link}", size}
-    end
+    build_table(remaining, ["<li>#{section_link}#{children}</li>" | acc])
```

Comment:
Truth be told I'd like to remove HTML from this part of the application all together. I'm going to take a stab at that before merging this.